packages/core/src/types.ts

```diff
@@ -52,34 +54,70 @@ export type UserOperationFeeOptions = z.input<
+> = Required extends true
+  ? { overrides: UserOperationOverrides<TEntryPointVersion> }
+  : { overrides?: UserOperationOverrides<TEntryPointVersion> };
+
 //#region UserOperationOverrides
-export type UserOperationOverrides = Partial<{
-  callGasLimit: UserOperationStruct["callGasLimit"] | Multiplier;
-  maxFeePerGas: UserOperationStruct["maxFeePerGas"] | Multiplier;
-  maxPriorityFeePerGas:
-    | UserOperationStruct["maxPriorityFeePerGas"]
-    | Multiplier;
-  preVerificationGas: UserOperationStruct["preVerificationGas"] | Multiplier;
-  verificationGasLimit:
-    | UserOperationStruct["verificationGasLimit"]
-    | Multiplier;
-  paymasterAndData: UserOperationStruct["paymasterAndData"];
-  /**
-   * This can be used to override the key used when calling `entryPoint.getNonce`
-   * It is useful when you want to use parallel nonces for user operations
-   *
-   * NOTE: not all bundlers fully support this feature and it could be that your bundler will still only include
-   * one user operation for your account in a bundle
-   */
-  nonceKey: bigint;
-  stateOverride: StateOverride;
-}>;
+// Note: entry point version is required for UserOperationOverrides
+export type UserOperationOverrides<
+  TEntryPointVersion extends EntryPointVersion
+> = Partial<
+  {
+    callGasLimit:
+      | UserOperationStruct<TEntryPointVersion>["callGasLimit"]
+      | Multiplier;
+    maxFeePerGas:
+      | UserOperationStruct<TEntryPointVersion>["maxFeePerGas"]
+      | Multiplier;
+    maxPriorityFeePerGas:
+      | UserOperationStruct<TEntryPointVersion>["maxPriorityFeePerGas"]
+      | Multiplier;
+    preVerificationGas:
+      | UserOperationStruct<TEntryPointVersion>["preVerificationGas"]
+      | Multiplier;
+    verificationGasLimit:
+      | UserOperationStruct<TEntryPointVersion>["verificationGasLimit"]
+      | Multiplier;
+    /**
+     * This can be used to override the key used when calling `entryPoint.getNonce`
+     * It is useful when you want to use parallel nonces for user operations
+     *
+     * NOTE: not all bundlers fully support this feature and it could be that your bundler will still only include
+     * one user operation for your account in a bundle
+     */
+    nonceKey: bigint;
+    stateOverride: StateOverride;
+  } & (TEntryPointVersion extends "0.6.0"
+    ? {
+        paymasterAndData: UserOperationStruct<"0.6.0">["paymasterAndData"];
+      }
+    : TEntryPointVersion extends "0.7.0"
+    ?
+        | {
+            paymaster: NoUndefined<UserOperationStruct<"0.7.0">["paymaster"]>;
+            paymasterData: NoUndefined<
+              UserOperationStruct<"0.7.0">["paymasterData"]
+            >;
+            paymasterPostOpGasLimit: NoUndefined<
+              UserOperationStruct<"0.7.0">["paymasterPostOpGasLimit"]
+            >;
+            paymasterVerificationGasLimit: NoUndefined<
+              UserOperationStruct<"0.7.0">["paymasterVerificationGasLimit"]
+            >;
+          }
+        | undefined
```

Comment:
yea we do this in A LOT of places, can we make this a type alias (think I left a comment above about this too)